File: conversion-system/voice/dograh_client.js
/**
 * Dograh API Client - Integration with Dograh AI Platform
 * Handles workflow creation, call initiation, and data retrieval
 */

const axios = require('axios');
const { EventEmitter } = require('events');
const Minio = require('minio');
const fs = require('fs');
const path = require('path');

class DograhClient extends EventEmitter {
    constructor(apiUrl = process.env.DOGRAH_API_URL || 'http://localhost:8000', apiKey = process.env.DOGRAH_API_KEY) {
        super();
        this.apiUrl = apiUrl;
        this.apiKey = apiKey;
        this.baseUrl = `${apiUrl}/api/v1`;

        // Initialize MinIO Client (for artifacts)
        this.minioClient = new Minio.Client({
            endPoint: process.env.MINIO_ENDPOINT || 'localhost',
            port: parseInt(process.env.MINIO_PORT || '9000'),
            useSSL: process.env.MINIO_USE_SSL === 'true',
            accessKey: process.env.MINIO_ACCESS_KEY || 'minioadmin',
            secretKey: process.env.MINIO_SECRET_KEY || 'minioadmin'
        });
        this.bucketName = process.env.MINIO_BUCKET || 'voice-audio';
    }

    /**
     * Get request headers with authentication
     */
    _getHeaders() {
        const headers = {
            'Content-Type': 'application/json'
        };

        if (this.apiKey) {
            const masked = this.apiKey.substring(0, 5) + '...' + this.apiKey.substring(this.apiKey.length - 5);
            console.log(`[Dograh] Auth Header: X-API-Key=${masked}`);
            headers['X-API-Key'] = this.apiKey.trim();
        }

        return headers;
    }

    // ... (healthCheck, createWorkflow, getWorkflow, listWorkflows, initiateCall methods remain unchanged) ...

    /**
     * Check Dograh API health
     */
    async healthCheck() {
        try {
            const response = await axios.get(`${this.apiUrl}/api/v1/health`);
            return response.data;
        } catch (error) {
            console.error('Dograh health check failed:', error.message);
            throw error;
        }
    }

    async createWorkflow(config) {
        try {
            const response = await axios.post(
                `${this.baseUrl}/workflows`,
                config,
                { headers: this._getHeaders() }
            );
            console.log(`[Dograh] Workflow created: ${response.data.id}`);
            return response.data;
        } catch (error) {
            console.error('[Dograh] Failed to create workflow:', error.message);
            throw error;
        }
    }

    async getWorkflow(workflowId) {
        try {
            const response = await axios.get(
                `${this.baseUrl}/workflows/${workflowId}`,
                { headers: this._getHeaders() }
            );
            return response.data;
        } catch (error) {
            console.error(`[Dograh] Failed to get workflow ${workflowId}:`, error.message);
            throw error;
        }
    }

    async listWorkflows() {
        try {
            const response = await axios.get(
                `${this.baseUrl}/workflows`,
                { headers: this._getHeaders() }
            );
            return response.data;
        } catch (error) {
            console.error('[Dograh] Failed to list workflows:', error.message);
            throw error;
        }
    }

    async initiateCall(triggerUuid, phoneNumber, context = {}) {
        try {
            // Preference: Use direct telephony initiation if workflow ID is available
            const workflowId = process.env.DOGRAH_WORKFLOW_ID;

            if (workflowId) {
                console.log(`[Dograh] Initiating call to ${phoneNumber} via direct Workflow ID ${workflowId}...`);
                const response = await axios.post(
                    `${this.baseUrl}/telephony/initiate-call`,
                    {
                        workflow_id: parseInt(workflowId),
                        phone_number: phoneNumber
                    },
                    {
                        headers: this._getHeaders(),
                        timeout: 60000
                    }
                );

                console.log(`[Dograh] Call initiation message: ${response.data.message}`);

                // The API returns an object like { runs: [], total_count: ... }.
                try {
                    const runsResponse = await axios.get(
                        `${this.baseUrl}/workflow/${workflowId}/runs`,
                        { headers: this._getHeaders() }
                    );

                    const latestRun = runsResponse.data.runs?.[0]; // Access the runs array
                    if (latestRun) {
                        console.log(`[Dograh] Identified Run ID: ${latestRun.id} (${latestRun.name})`);
                        return {
                            call_id: latestRun.id,
                            ...latestRun
                        };
                    }
                } catch (runError) {
                    console.warn('[Dograh] Failed to fetch run ID, using fallback string parsing', runError.message);
                }

                return {
                    call_id: response.data.message?.split(' ').pop() || 'unknown',
                    ...response.data
                };
            }

            // Fallback: Public trigger UUID method
            const payload = {
                phone_number: phoneNumber,
                initial_context: context
            };

            console.log(`[Dograh] Initiating call to ${phoneNumber} with trigger ${triggerUuid}`);

            const response = await axios.post(
                `${this.baseUrl}/public/agent/${triggerUuid}`,
                payload,
                {
                    headers: this._getHeaders(),
                    timeout: 60000
                }
            );

            console.log(`[Dograh] Call initiated successfully`);
            return {
                call_id: response.data.workflow_run_id,
                ...response.data
            };
        } catch (error) {
            console.error('[Dograh] Failed to initiate call:', error.response?.data || error.message);
            throw error;
        }
    }

    async getCallStatus(workflowRunId, workflowId) {
        try {
            const response = await axios.get(
                `${this.baseUrl}/workflow/${workflowId}/runs/${workflowRunId}`,
                { headers: this._getHeaders() }
            );
            const run = response.data;
            return {
                status: run.is_completed ? 'completed' : (run.status || 'in-progress'),
                duration: run.cost_info?.call_duration_seconds || 0,
                workflow_run_id: run.id,
                ...run
            };
        } catch (error) {
            console.error(`[Dograh] Failed to get status for run ${workflowRunId}:`, error.message);
            throw error;
        }
    }

    /**
     * Download artifact (transcript/recording) from MinIO
     */
    async downloadArtifact(runId, type) {
        const ext = type === 'transcript' ? 'txt' : 'wav';
        const folder = type === 'transcript' ? 'transcripts' : 'recordings';
        const objectName = `${folder}/${runId}.${ext}`;

        // Ensure local directory exists
        const localDir = path.join(__dirname, `../voice/${folder}`);
        if (!fs.existsSync(localDir)) fs.mkdirSync(localDir, { recursive: true });

        const localPath = path.join(localDir, `${runId}.${ext}`);

        try {
            await this.minioClient.fGetObject(this.bucketName, objectName, localPath);
            // console.log(`[Dograh] Downloaded ${type} to ${localPath}`);
            return localPath;
        } catch (err) {
            console.warn(`[Dograh] Failed to download ${type} for run ${runId}: ${err.message}`);
            return null;
        }
    }

    /**
     * Get call transcript from workflow run
     */
    async getCallTranscript(workflowRunId, workflowId) {
        try {
            const response = await axios.get(
                `${this.baseUrl}/workflow/${workflowId}/runs/${workflowRunId}`,
                { headers: this._getHeaders() }
            );

            const run = response.data;
            let transcript = '';
            let messages = [];

            // 1. Try to get from logs 
            if (run.logs) {
                if (run.logs.transcript) transcript = run.logs.transcript;
                if (run.logs.messages) messages = run.logs.messages;
            }

            // 2. If no logs, try downloading from MinIO
            if (!transcript && run.transcript_url) {
                const localPath = await this.downloadArtifact(workflowRunId, 'transcript');
                if (localPath && fs.existsSync(localPath)) {
                    transcript = fs.readFileSync(localPath, 'utf8');
                }
            }

            // Always try to download recording too
            if (run.recording_url) {
                this.downloadArtifact(workflowRunId, 'recording').catch(() => { });
            }

            return {
                transcript: transcript || 'No transcript available',
                summary: run.logs?.summary || '',
                messages: messages
            };
        } catch (error) {
            console.error(`[Dograh] Failed to get transcript for run ${workflowRunId}:`, error.message);
            throw error;
        }
    }

    async getCallVariables(workflowRunId, workflowId) {
        try {
            const response = await axios.get(
                `${this.baseUrl}/workflow/${workflowId}/runs/${workflowRunId}`,
                { headers: this._getHeaders() }
            );
            const run = response.data;
            return run.gathered_context || run.initial_context || {};
        } catch (error) {
            console.error(`[Dograh] Failed to get variables for run ${workflowRunId}:`, error.message);
            throw error;
        }
    }

    async getCallRecording(callId) {
        // Deprecated in favor of direct download, but kept for compatibility
        return this.downloadArtifact(callId, 'recording');
    }

    /**
     * Wait for call completion and get final data
     * @param {string} callId - Call ID
     * @param {string} workflowId - Workflow ID
     * @param {number} maxWaitMs - Maximum wait time in milliseconds
     */
    async waitForCallCompletion(callId, workflowId, maxWaitMs = 600000) {
        const startTime = Date.now();
        const pollInterval = 5000; // Poll every 5 seconds

        while (Date.now() - startTime < maxWaitMs) {
            try {
                const status = await this.getCallStatus(callId, workflowId);

                // Check if call is in terminal state
                if (status.is_completed || ['completed', 'failed', 'no-answer', 'busy'].includes(status.status)) {
                    console.log(`[Dograh] Call ${callId} completed with status: ${status.status}`);

                    // Get final data
                    const [transcript, variables] = await Promise.all([
                        this.getCallTranscript(callId, workflowId).catch(() => null),
                        this.getCallVariables(callId, workflowId).catch(() => null)
                    ]);

                    return {
                        status: status.status,
                        transcript,
                        variables,
                        duration: status.duration,
                        call_id: callId
                    };
                }

                // Wait before next poll
                await new Promise(resolve => setTimeout(resolve, pollInterval));
            } catch (error) {
                console.error(`[Dograh] Error polling call status:`, error.message);
                await new Promise(resolve => setTimeout(resolve, pollInterval));
            }
        }

        throw new Error(`Call ${callId} did not complete within ${maxWaitMs}ms`);
    }
}

module.exports = DograhClient;

File: conversion-system/python-system/voice/dograh_client.py
# ---------------------------------------------------------
# voice/dograh_client.py
# Dograh AI Platform client - replaces voice/dograh_client.js
# ---------------------------------------------------------
from __future__ import annotations

import asyncio
from pathlib import Path
from typing import Any

import httpx

from core.config import (
    DOGRAH_API_KEY,
    DOGRAH_API_URL,
    DOGRAH_WORKFLOW_ID,
    MINIO_ACCESS_KEY,
    MINIO_BUCKET,
    MINIO_ENDPOINT,
    MINIO_PORT,
    MINIO_SECRET_KEY,
    MINIO_USE_SSL,
)


class DograhClient:
    """
    HTTP client for the Dograh AI Platform.
    Equivalent to DograhClient class in voice/dograh_client.js
    """

    def __init__(
        self,
        api_url: str = DOGRAH_API_URL,
        api_key: str = DOGRAH_API_KEY,
    ):
        self.api_url = api_url.rstrip("/")
        self.api_key = api_key
        self.base_url = f"{self.api_url}/api/v1"

    def _headers(self) -> dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            masked = self.api_key[:5] + "..." + self.api_key[-5:]
            print(f"[Dograh] Auth Header: X-API-Key={masked}")
            headers["X-API-Key"] = self.api_key.strip()
        return headers

    async def health_check(self) -> dict:
        async with httpx.AsyncClient(timeout=10) as client:
            r = await client.get(f"{self.api_url}/api/v1/health")
            return r.json()

    async def create_workflow(self, config: dict) -> dict:
        async with httpx.AsyncClient(timeout=30) as client:
            r = await client.post(f"{self.base_url}/workflows", json=config, headers=self._headers())
        print(f"[Dograh] Workflow created: {r.json().get('id')}")
        return r.json()

    async def get_workflow(self, workflow_id: str) -> dict:
        async with httpx.AsyncClient(timeout=10) as client:
            r = await client.get(f"{self.base_url}/workflows/{workflow_id}", headers=self._headers())
        return r.json()

    async def list_workflows(self) -> dict:
        async with httpx.AsyncClient(timeout=10) as client:
            r = await client.get(f"{self.base_url}/workflows", headers=self._headers())
        return r.json()

    async def initiate_call(
        self,
        trigger_uuid: str,
        phone_number: str,
        context: dict | None = None,
    ) -> dict[str, Any]:
        """
        Initiates a voice call via Dograh.
        Equivalent to initiateCall() in dograh_client.js
        """
        context = context or {}
        workflow_id = DOGRAH_WORKFLOW_ID

        if workflow_id:
            print(f"[Dograh] Initiating call to {phone_number} via Workflow ID {workflow_id}...")
            async with httpx.AsyncClient(timeout=60) as client:
                r = await client.post(
                    f"{self.base_url}/telephony/initiate-call",
                    json={"workflow_id": int(workflow_id), "phone_number": phone_number},
                    headers=self._headers(),
                )
            data = r.json()
            print(f"[Dograh] Call initiation message: {data.get('message')}")

            # Try to fetch the resulting run ID
            try:
                async with httpx.AsyncClient(timeout=10) as client:
                    runs_r = await client.get(
                        f"{self.base_url}/workflow/{workflow_id}/runs",
                        headers=self._headers(),
                    )
                runs = runs_r.json().get("runs", [])
                latest = runs[0] if runs else None
                if latest:
                    print(f"[Dograh] Identified Run ID: {latest['id']} ({latest.get('name')})")
                    return {"call_id": latest["id"], **latest}
            except Exception as e:
                print(f"[Dograh] Failed to fetch run ID: {e}")

            return {"call_id": str(data.get("message", "")).split()[-1] or "unknown", **data}

        # Fallback: public trigger UUID method
        print(f"[Dograh] Initiating call to {phone_number} with trigger {trigger_uuid}")
        payload = {"phone_number": phone_number, "initial_context": context}
        async with httpx.AsyncClient(timeout=60) as client:
            r = await client.post(
                f"{self.base_url}/public/agent/{trigger_uuid}",
                json=payload,
                headers=self._headers(),
            )
        data = r.json()
        print("[Dograh] Call initiated successfully")
        return {"call_id": data.get("workflow_run_id"), **data}

    async def get_call_status(self, workflow_run_id: str, workflow_id: str) -> dict:
        async with httpx.AsyncClient(timeout=10) as client:
            r = await client.get(
                f"{self.base_url}/workflow/{workflow_id}/runs/{workflow_run_id}",
                headers=self._headers(),
            )
        run = r.json()
        return {
            "status": "completed" if run.get("is_completed") else run.get("status", "in-progress"),
            "duration": (run.get("cost_info") or {}).get("call_duration_seconds", 0),
            "workflow_run_id": run.get("id"),
            **run,
        }

    async def get_call_transcript(self, workflow_run_id: str, workflow_id: str) -> dict:
        async with httpx.AsyncClient(timeout=10) as client:
            r = await client.get(
                f"{self.base_url}/workflow/{workflow_id}/runs/{workflow_run_id}",
                headers=self._headers(),
            )
        run = r.json()
        transcript = ""
        messages = []
        logs = run.get("logs") or {}
        if logs.get("transcript"):
            transcript = logs["transcript"]
        if logs.get("messages"):
            messages = logs["messages"]

        return {
            "transcript": transcript or "No transcript available",
            "summary": logs.get("summary", ""),
            "messages": messages,
        }

    async def get_call_variables(self, workflow_run_id: str, workflow_id: str) -> dict:
        async with httpx.AsyncClient(timeout=10) as client:
            r = await client.get(
                f"{self.base_url}/workflow/{workflow_id}/runs/{workflow_run_id}",
                headers=self._headers(),
            )
        run = r.json()
        return run.get("gathered_context") or run.get("initial_context") or {}

    async def wait_for_call_completion(
        self,
        call_id: str,
        workflow_id: str,
        max_wait_ms: int = 600_000,
    ) -> dict[str, Any]:
        """
        Polls call status until terminal state or timeout.
        Equivalent to waitForCallCompletion() in dograh_client.js
        """
        import time

        start = time.time()
        poll_interval = 5  # seconds

        while (time.time() - start) * 1000 < max_wait_ms:
            try:
                status = await self.get_call_status(call_id, workflow_id)
                terminal = {"completed", "failed", "no-answer", "busy"}
                if status.get("is_completed") or status.get("status") in terminal:
                    print(f"[Dograh] Call {call_id} completed with status: {status.get('status')}")

                    transcript, variables = await asyncio.gather(
                        self.get_call_transcript(call_id, workflow_id),
                        self.get_call_variables(call_id, workflow_id),
                        return_exceptions=True,
                    )
                    if isinstance(transcript, Exception):
                        transcript = None
                    if isinstance(variables, Exception):
                        variables = None

                    return {
                        "status": status.get("status"),
                        "transcript": transcript,
                        "variables": variables,
                        "duration": status.get("duration"),
                        "call_id": call_id,
                    }
            except Exception as error:
                print(f"[Dograh] Error polling call status: {error}")

            await asyncio.sleep(poll_interval)

        raise TimeoutError(f"Call {call_id} did not complete within {max_wait_ms}ms")
